std::unordered_map conformance to CxxDictionary does not work on Linux

```
******************** TEST 'Swift(linux-x86_64) :: Interop/Cxx/stdlib/use-std-map.swift' FAILED ********************
Script:
--
: 'RUN: at line 1';   rm -rf "/home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Interop/Cxx/stdlib/Output/use-std-map.swift.tmp" && mkdir -p "/home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Interop/Cxx/stdlib/Output/use-std-map.swift.tmp" && /home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swiftc -target x86_64-unknown-linux-gnu -toolchain-stdlib-rpath  -module-cache-path /home/build-user/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -swift-version 4  -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 9999:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.0:macOS 10.14.4, iOS 12.2, watchOS 5.2, tvOS 12.2' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.1:macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.2:macOS 10.15.4, iOS 13.4, watchOS 6.2, tvOS 13.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.3:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.4:macOS 11.3, iOS 14.5, watchOS 7.4, tvOS 14.5' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.5:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.6:macOS 12.3, iOS 15.4, watchOS 8.5, tvOS 15.4' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.7:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.8:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999' -Xfrontend -define-availability -Xfrontend 'SwiftStdlib 5.9:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999'  -module-cache-path /home/build-user/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache /home/build-user/swift/test/Interop/Cxx/stdlib/use-std-map.swift -I /home/build-user/swift/test/Interop/Cxx/stdlib/Inputs -Xfrontend -enable-experimental-cxx-interop -Xfrontend -validate-tbd-against-ir=none -o /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Interop/Cxx/stdlib/Output/use-std-map.swift.tmp/a.out -module-name main  && echo /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Interop/Cxx/stdlib/Output/use-std-map.swift.tmp/a.out && /usr/bin/env LD_LIBRARY_PATH='/home/build-user/build/buildbot_linux/swift-linux-x86_64/lib/swift/linux'  /home/build-user/build/buildbot_linux/swift-linux-x86_64/test-linux-x86_64/Interop/Cxx/stdlib/Output/use-std-map.swift.tmp/a.out
--
Exit Code: 254

Command Output (stderr):
--
/home/build-user/swift/test/Interop/Cxx/stdlib/use-std-map.swift:25:7: warning: variable 'm' was never mutated; consider changing to 'let' constant
  var m = initMap()
  ~~~ ^
  let
/home/build-user/swift/test/Interop/Cxx/stdlib/use-std-map.swift:37:7: warning: variable 'm' was never mutated; consider changing to 'let' constant
  var m = initUnorderedMap()
  ~~~ ^
  let
<unknown>:0: error: fatal error encountered during compilation; please submit a bug report (https://swift.org/contributing/#reporting-bugs)
<unknown>:0: note: Invalid conformance in type-checked AST
Please submit a bug report (https://swift.org/contributing/#reporting-bugs) and include the crash backtrace.
Stack dump:
0.	Program arguments: /home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend -frontend -c -primary-file /home/build-user/swift/test/Interop/Cxx/stdlib/use-std-map.swift -target x86_64-unknown-linux-gnu -disable-objc-interop -I /home/build-user/swift/test/Interop/Cxx/stdlib/Inputs -module-cache-path /home/build-user/build/buildbot_linux/swift-linux-x86_64/swift-test-results/x86_64-unknown-linux-gnu/clang-module-cache -swift-version 4 -define-availability "SwiftStdlib 9999:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" -define-availability "SwiftStdlib 5.0:macOS 10.14.4, iOS 12.2, watchOS 5.2, tvOS 12.2" -define-availability "SwiftStdlib 5.1:macOS 10.15, iOS 13.0, watchOS 6.0, tvOS 13.0" -define-availability "SwiftStdlib 5.2:macOS 10.15.4, iOS 13.4, watchOS 6.2, tvOS 13.4" -define-availability "SwiftStdlib 5.3:macOS 11.0, iOS 14.0, watchOS 7.0, tvOS 14.0" -define-availability "SwiftStdlib 5.4:macOS 11.3, iOS 14.5, watchOS 7.4, tvOS 14.5" -define-availability "SwiftStdlib 5.5:macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0" -define-availability "SwiftStdlib 5.6:macOS 12.3, iOS 15.4, watchOS 8.5, tvOS 15.4" -define-availability "SwiftStdlib 5.7:macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0" -define-availability "SwiftStdlib 5.8:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" -define-availability "SwiftStdlib 5.9:macOS 9999, iOS 9999, watchOS 9999, tvOS 9999" -enable-experimental-cxx-interop -validate-tbd-against-ir=none -module-name main -o /tmp/lit-tmp-3zm5rzp8/use-std-map-4b9bed.o
1.	Swift version 5.9-dev (LLVM 5ea1e8195110c60, Swift 2eb668557eb5a44)
2.	Compiling with effective version 4.1.50
3.	While evaluating request ASTLoweringRequest(Lowering AST to SIL for file "/home/build-user/swift/test/Interop/Cxx/stdlib/use-std-map.swift")
4.	While generating SIL witness table protocol conformance to 'CxxDictionary' (in module 'Cxx') for type 'std.__CxxTemplateInstSt13unordered_mapIiiSt4hashIiESt8equal_toIiESaISt4pairIKiiEEE'
 #0 0x000000000661c843 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x661c843)
 #1 0x000000000661a57e llvm::sys::RunSignalHandlers() (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x661a57e)
 #2 0x000000000661cbcf SignalHandler(int) Signals.cpp:0:0
 #3 0x00007ff410dab420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #4 0x00007ff41021c00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #5 0x00007ff4101fb859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #6 0x00000000006a10ff swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*)::$_3::__invoke(void*, char const*, bool) FrontendTool.cpp:0:0
 #7 0x0000000006542ff9 llvm::report_fatal_error(llvm::Twine const&, bool) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x6542ff9)
 #8 0x0000000006542ed8 (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x6542ed8)
 #9 0x0000000000d8433d swift::Lowering::SILGenModule::useConformance(swift::ProtocolConformanceRef) crtstuff.c:0:0
#10 0x0000000000dd9e4a swift::SILWitnessVisitor<(anonymous namespace)::SILGenConformance>::visitProtocolDecl(swift::ProtocolDecl*) SILGenType.cpp:0:0
#11 0x0000000000dd70fc swift::Lowering::SILGenModule::getWitnessTable(swift::NormalProtocolConformance*) crtstuff.c:0:0
#12 0x0000000000d07445 swift::ASTLoweringRequest::evaluate(swift::Evaluator&, swift::ASTLoweringDescriptor) const (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0xd07445)
#13 0x0000000000dca30f swift::SimpleRequest<swift::ASTLoweringRequest, std::unique_ptr<swift::SILModule, std::default_delete<swift::SILModule> > (swift::ASTLoweringDescriptor), (swift::RequestFlags)9>::evaluateRequest(swift::ASTLoweringRequest const&, swift::Evaluator&) crtstuff.c:0:0
#14 0x0000000000d0aada llvm::Expected<swift::ASTLoweringRequest::OutputType> swift::Evaluator::getResultUncached<swift::ASTLoweringRequest>(swift::ASTLoweringRequest const&) crtstuff.c:0:0
#15 0x0000000000d079f9 swift::performASTLowering(swift::FileUnit&, swift::Lowering::TypeConverter&, swift::SILOptions const&, swift::IRGenOptions const*) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0xd079f9)
#16 0x000000000069577d swift::performCompileStepsPostSema(swift::CompilerInstance&, int&, swift::FrontendObserver*) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x69577d)
#17 0x000000000069882c swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x69882c)
#18 0x00000000004c03b9 swift::mainEntry(int, char const**) (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x4c03b9)
#19 0x00007ff4101fd083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#20 0x00000000004bfb1e _start (/home/build-user/build/buildbot_linux/swift-linux-x86_64/bin/swift-frontend+0x4bfb1e)
<unknown>:0: error: unable to execute command: Aborted
<unknown>:0: error: compile command failed due to signal 6 (use -v to see invocation)

--

********************
```

--- a/include/swift/ClangImporter/ClangImporter.h
+++ b/include/swift/ClangImporter/ClangImporter.h
@@ -568,6 +568,8 @@ public:
                                  clang::FunctionTemplateDecl *func,
                                  SubstitutionMap subst) override;
 
+  bool isSynthesizedAndVisibleFromAllModules(const clang::Decl *decl);
+
   bool isCXXMethodMutating(const clang::CXXMethodDecl *method) override;
 
   bool isUnsafeCXXMethod(const FuncDecl *func) override;

--- a/lib/ClangImporter/ClangDerivedConformances.cpp
+++ b/lib/ClangImporter/ClangDerivedConformances.cpp
@@ -156,7 +156,7 @@ static ValueDecl *getEqualEqualOperator(NominalTypeDecl *decl) {
   return lookupOperator(decl, decl->getASTContext().Id_EqualsOperator, isValid);
 }
 
-static ValueDecl *getMinusOperator(NominalTypeDecl *decl) {
+static FuncDecl *getMinusOperator(NominalTypeDecl *decl) {
   auto binaryIntegerProto =
       decl->getASTContext().getProtocol(KnownProtocolKind::BinaryInteger);
   auto module = decl->getModuleContext();
@@ -188,11 +188,12 @@ static ValueDecl *getMinusOperator(NominalTypeDecl *decl) {
     return true;
   };
 
-  return lookupOperator(decl, decl->getASTContext().getIdentifier("-"),
-                        isValid);
+  ValueDecl *result =
+      lookupOperator(decl, decl->getASTContext().getIdentifier("-"), isValid);
+  return dyn_cast_or_null<FuncDecl>(result);
 }
 
-static ValueDecl *getPlusEqualOperator(NominalTypeDecl *decl, Type distanceTy) {
+static FuncDecl *getPlusEqualOperator(NominalTypeDecl *decl, Type distanceTy) {
   auto isValid = [&](ValueDecl *plusEqualOp) -> bool {
     auto plusEqual = dyn_cast<FuncDecl>(plusEqualOp);
     if (!plusEqual || !plusEqual->hasParameterList())
@@ -219,14 +220,15 @@ static ValueDecl *getPlusEqualOperator(NominalTypeDecl *decl, Type distanceTy) {
     return true;
   };
 
-  return lookupOperator(decl, decl->getASTContext().getIdentifier("+="),
-                        isValid);
+  ValueDecl *result =
+      lookupOperator(decl, decl->getASTContext().getIdentifier("+="), isValid);
+  return dyn_cast_or_null<FuncDecl>(result);
 }
 
-static void instantiateTemplatedOperator(
-    ClangImporter::Implementation &impl,
-    const clang::ClassTemplateSpecializationDecl *classDecl,
-    clang::BinaryOperatorKind operatorKind) {
+static clang::FunctionDecl *
+instantiateTemplatedOperator(ClangImporter::Implementation &impl,
+                             const clang::CXXRecordDecl *classDecl,
+                             clang::BinaryOperatorKind operatorKind) {
 
   clang::ASTContext &clangCtx = impl.getClangASTContext();
   clang::Sema &clangSema = impl.getClangSema();
@@ -252,6 +254,7 @@ static void instantiateTemplatedOperator(
     if (auto clangCallee = best->Function) {
       auto lookupTable = impl.findLookupTable(classDecl);
       addEntryToLookupTable(*lookupTable, clangCallee, impl.getNameImporter());
+      return clangCallee;
     }
     break;
   }
@@ -260,6 +263,95 @@ static void instantiateTemplatedOperator(
   case clang::OR_Deleted:
     break;
   }
+
+  return nullptr;
+}
+
+/// Warning: This function emits an error and stops compilation if the
+/// underlying operator function is unavailable in Swift for the current target
+/// (see `clang::Sema::DiagnoseAvailabilityOfDecl`).
+static bool synthesizeCXXOperator(ClangImporter::Implementation &impl,
+                                  const clang::CXXRecordDecl *classDecl,
+                                  clang::BinaryOperatorKind operatorKind,
+                                  clang::QualType lhsTy, clang::QualType rhsTy,
+                                  clang::QualType returnTy) {
+  auto &clangCtx = impl.getClangASTContext();
+  auto &clangSema = impl.getClangSema();
+
+  clang::OverloadedOperatorKind opKind =
+      clang::BinaryOperator::getOverloadedOperator(operatorKind);
+  const char *opSpelling = clang::getOperatorSpelling(opKind);
+
+  auto declName = clang::DeclarationName(&clangCtx.Idents.get(opSpelling));
+
+  // Determine the Clang decl context where the new operator function will be
+  // created. We use the translation unit as the decl context of the new
+  // operator, otherwise, the operator might get imported as a static member
+  // function of a different type (e.g. an operator declared inside of a C++
+  // namespace would get imported as a member function of a Swift enum), which
+  // would make the operator un-discoverable to Swift name lookup.
+  auto declContext =
+      const_cast<clang::CXXRecordDecl *>(classDecl)->getDeclContext();
+  while (!declContext->isTranslationUnit()) {
+    declContext = declContext->getParent();
+  }
+
+  auto equalEqualTy = clangCtx.getFunctionType(
+      returnTy, {lhsTy, rhsTy}, clang::FunctionProtoType::ExtProtoInfo());
+
+  // Create a `bool operator==(T, T)` function.
+  auto equalEqualDecl = clang::FunctionDecl::Create(
+      clangCtx, declContext, clang::SourceLocation(), clang::SourceLocation(),
+      declName, equalEqualTy, clangCtx.getTrivialTypeSourceInfo(returnTy),
+      clang::StorageClass::SC_Static);
+  equalEqualDecl->setImplicit();
+  equalEqualDecl->setImplicitlyInline();
+  // If this is a static member function of a class, it needs to be public.
+  equalEqualDecl->setAccess(clang::AccessSpecifier::AS_public);
+
+  // Create the parameters of the function. They are not referenced from source
+  // code, so they don't need to have a name.
+  auto lhsParamId = nullptr;
+  auto lhsTyInfo = clangCtx.getTrivialTypeSourceInfo(lhsTy);
+  auto lhsParamDecl = clang::ParmVarDecl::Create(
+      clangCtx, equalEqualDecl, clang::SourceLocation(),
+      clang::SourceLocation(), lhsParamId, lhsTy, lhsTyInfo,
+      clang::StorageClass::SC_None, /*DefArg*/ nullptr);
+  auto lhsParamRefExpr = new (clangCtx) clang::DeclRefExpr(
+      clangCtx, lhsParamDecl, false, lhsTy, clang::ExprValueKind::VK_LValue,
+      clang::SourceLocation());
+
+  auto rhsParamId = nullptr;
+  auto rhsTyInfo = clangCtx.getTrivialTypeSourceInfo(rhsTy);
+  auto rhsParamDecl = clang::ParmVarDecl::Create(
+      clangCtx, equalEqualDecl, clang::SourceLocation(),
+      clang::SourceLocation(), rhsParamId, rhsTy, rhsTyInfo,
+      clang::StorageClass::SC_None, nullptr);
+  auto rhsParamRefExpr = new (clangCtx) clang::DeclRefExpr(
+      clangCtx, rhsParamDecl, false, rhsTy, clang::ExprValueKind::VK_LValue,
+      clang::SourceLocation());
+
+  equalEqualDecl->setParams({lhsParamDecl, rhsParamDecl});
+
+  // Lookup the `operator==` function that will be called under the hood.
+  clang::UnresolvedSet<16> operators;
+  // Note: calling `CreateOverloadedBinOp` emits an error if the looked up
+  // function is unavailable for the current target.
+  auto underlyingCallResult = clangSema.CreateOverloadedBinOp(
+      clang::SourceLocation(), operatorKind, operators, lhsParamRefExpr,
+      rhsParamRefExpr);
+  if (!underlyingCallResult.isUsable())
+    return false;
+  auto underlyingCall = underlyingCallResult.get();
+
+  auto equalEqualBody = clang::ReturnStmt::Create(
+      clangCtx, clang::SourceLocation(), underlyingCall, nullptr);
+  equalEqualDecl->setBody(equalEqualBody);
+
+  impl.synthesizedAndAlwaysVisibleDecls.insert(equalEqualDecl);
+  auto lookupTable = impl.findLookupTable(classDecl);
+  addEntryToLookupTable(*lookupTable, equalEqualDecl, impl.getNameImporter());
+  return true;
 }
 
 bool swift::isIterator(const clang::CXXRecordDecl *clangDecl) {
@@ -274,6 +366,7 @@ void swift::conformToCxxIteratorIfNeeded(
   assert(decl);
   assert(clangDecl);
   ASTContext &ctx = decl->getASTContext();
+  clang::ASTContext &clangCtx = clangDecl->getASTContext();
 
   if (!ctx.getProtocol(KnownProtocolKind::UnsafeCxxInputIterator))
     return;
@@ -349,15 +442,28 @@ void swift::conformToCxxIteratorIfNeeded(
   if (!successorTy || successorTy->getAnyNominal() != decl)
     return;
 
-  // If this is a templated class, `operator==` might be templated as well.
-  // Try to instantiate it.
-  if (auto templateSpec =
-          dyn_cast<clang::ClassTemplateSpecializationDecl>(clangDecl)) {
-    instantiateTemplatedOperator(impl, templateSpec,
-                                 clang::BinaryOperatorKind::BO_EQ);
-  }
   // Check if present: `func ==`
   auto equalEqual = getEqualEqualOperator(decl);
+  if (!equalEqual) {
+    // If this class is inherited, `operator==` might be defined for a base
+    // class. If this is a templated class, `operator==` might be templated as
+    // well. Try to instantiate it.
+    clang::FunctionDecl *instantiated = instantiateTemplatedOperator(
+        impl, clangDecl, clang::BinaryOperatorKind::BO_EQ);
+    if (instantiated && !impl.isUnavailableInSwift(instantiated)) {
+      // If `operator==` was instantiated successfully, try to find `func ==`
+      // again.
+      equalEqual = getEqualEqualOperator(decl);
+      if (!equalEqual) {
+        // If `func ==` still can't be found, it might be defined for a base
+        // class of the current class.
+        auto paramTy = clangCtx.getRecordType(clangDecl);
+        synthesizeCXXOperator(impl, clangDecl, clang::BinaryOperatorKind::BO_EQ,
+                              paramTy, paramTy, clangCtx.BoolTy);
+        equalEqual = getEqualEqualOperator(decl);
+      }
+    }
+  }
   if (!equalEqual)
     return;
 
@@ -371,18 +477,46 @@ void swift::conformToCxxIteratorIfNeeded(
 
   // Try to conform to UnsafeCxxRandomAccessIterator if possible.
 
-  if (auto templateSpec =
-          dyn_cast<clang::ClassTemplateSpecializationDecl>(clangDecl)) {
-    instantiateTemplatedOperator(impl, templateSpec,
-                                 clang::BinaryOperatorKind::BO_Sub);
+  // Check if present: `func -`
+  auto minus = getMinusOperator(decl);
+  if (!minus) {
+    clang::FunctionDecl *instantiated = instantiateTemplatedOperator(
+        impl, clangDecl, clang::BinaryOperatorKind::BO_Sub);
+    if (instantiated && !impl.isUnavailableInSwift(instantiated)) {
+      minus = getMinusOperator(decl);
+      if (!minus) {
+        clang::QualType returnTy = instantiated->getReturnType();
+        auto paramTy = clangCtx.getRecordType(clangDecl);
+        synthesizeCXXOperator(impl, clangDecl,
+                              clang::BinaryOperatorKind::BO_Sub, paramTy,
+                              paramTy, returnTy);
+        minus = getMinusOperator(decl);
+      }
+    }
   }
-  auto minus = dyn_cast_or_null<FuncDecl>(getMinusOperator(decl));
   if (!minus)
     return;
   auto distanceTy = minus->getResultInterfaceType();
   // distanceTy conforms to BinaryInteger, this is ensured by getMinusOperator.
 
-  auto plusEqual = dyn_cast_or_null<FuncDecl>(getPlusEqualOperator(decl, distanceTy));
+  auto plusEqual = getPlusEqualOperator(decl, distanceTy);
+  if (!plusEqual) {
+    clang::FunctionDecl *instantiated = instantiateTemplatedOperator(
+        impl, clangDecl, clang::BinaryOperatorKind::BO_AddAssign);
+    if (instantiated && !impl.isUnavailableInSwift(instantiated)) {
+      plusEqual = getPlusEqualOperator(decl, distanceTy);
+      if (!plusEqual) {
+        clang::QualType returnTy = instantiated->getReturnType();
+        auto clangMinus = cast<clang::FunctionDecl>(minus->getClangDecl());
+        auto lhsTy = clangCtx.getRecordType(clangDecl);
+        auto rhsTy = clangMinus->getReturnType();
+        synthesizeCXXOperator(impl, clangDecl,
+                              clang::BinaryOperatorKind::BO_AddAssign, lhsTy,
+                              rhsTy, returnTy);
+        plusEqual = getPlusEqualOperator(decl, distanceTy);
+      }
+    }
+  }
   if (!plusEqual)
     return;
 

--- a/lib/ClangImporter/ClangImporter.cpp
+++ b/lib/ClangImporter/ClangImporter.cpp
@@ -2842,6 +2842,12 @@ static bool isVisibleFromModule(const ClangModuleUnit *ModuleFilter,
   if (OwningClangModule == ModuleFilter->getClangModule())
     return true;
 
+  // If this decl was implicitly synthesized by the compiler, and is not
+  // supposed to be owned by any module, return true.
+  if (Importer->isSynthesizedAndVisibleFromAllModules(D)) {
+    return true;
+  }
+
   // Friends from class templates don't have an owning module. Just return true.
   if (isa<clang::FunctionDecl>(D) &&
       cast<clang::FunctionDecl>(D)->isThisDeclarationInstantiatedFromAFriendDefinition())
@@ -6300,6 +6306,11 @@ FuncDecl *ClangImporter::getCXXSynthesizedOperatorFunc(FuncDecl *decl) {
   assert(synthesizedOperator->isOperator() &&
          "expected the alternative to be a synthesized operator");
   return cast<FuncDecl>(synthesizedOperator);
+}
+
+bool ClangImporter::isSynthesizedAndVisibleFromAllModules(
+    const clang::Decl *decl) {
+  return Impl.synthesizedAndAlwaysVisibleDecls.contains(decl);
 }
 
 bool ClangImporter::isCXXMethodMutating(const clang::CXXMethodDecl *method) {

--- a/lib/ClangImporter/ImporterImpl.h
+++ b/lib/ClangImporter/ImporterImpl.h
@@ -642,6 +642,8 @@ public:
   llvm::MapVector<std::pair<NominalTypeDecl *, Type>,
                   std::pair<FuncDecl *, FuncDecl *>> cxxSubscripts;
 
+  llvm::SmallPtrSet<const clang::Decl *, 1> synthesizedAndAlwaysVisibleDecls;
+
 private:
   // Keep track of the decls that were already cloned for this specific class.
   llvm::DenseMap<std::pair<ValueDecl *, DeclContext *>, ValueDecl *>

--- a/test/Interop/Cxx/stdlib/overlay/Inputs/custom-collection.h
+++ b/test/Interop/Cxx/stdlib/overlay/Inputs/custom-collection.h
@@ -28,4 +28,38 @@ public:
   const int& operator[](int index) const { return x[index]; }
 };
 
+template <typename T>
+struct HasInheritedTemplatedConstRACIterator {
+public:
+  typedef InheritedTemplatedConstRACIterator<int> iterator;
+
+private:
+  iterator b = iterator(1);
+  iterator e = iterator(6);
+
+public:
+  iterator begin() const { return b; }
+  iterator end() const { return e; }
+};
+
+typedef HasInheritedTemplatedConstRACIterator<int>
+    HasInheritedTemplatedConstRACIteratorInt;
+
+template <typename T>
+struct HasInheritedTemplatedConstRACIteratorOutOfLineOps {
+public:
+  typedef InheritedTemplatedConstRACIteratorOutOfLineOps<int> iterator;
+
+private:
+  iterator b = iterator(1);
+  iterator e = iterator(4);
+
+public:
+  iterator begin() const { return b; }
+  iterator end() const { return e; }
+};
+
+typedef HasInheritedTemplatedConstRACIteratorOutOfLineOps<int>
+    HasInheritedTemplatedConstRACIteratorOutOfLineOpsInt;
+
 #endif // TEST_INTEROP_CXX_STDLIB_INPUTS_CUSTOM_COLLECTION_H

--- a/test/Interop/Cxx/stdlib/overlay/Inputs/custom-iterator.h
+++ b/test/Interop/Cxx/stdlib/overlay/Inputs/custom-iterator.h
@@ -593,4 +593,241 @@ operator-(const TemplatedRACIteratorOutOfLineEq<T> &lhs,
 
 using TemplatedRACIteratorOutOfLineEqInt = TemplatedRACIteratorOutOfLineEq<int>;
 
+// MARK: Iterator types that use inheritance
+
+struct BaseIntIterator {
+  using value_type = int;
+  using reference = const int &;
+
+  int value;
+
+  BaseIntIterator(int value) : value(value) {}
+
+  reference operator*() const { return value; }
+
+  bool operator==(const BaseIntIterator &other) const {
+    return value == other.value;
+  }
+  bool operator!=(const BaseIntIterator &other) const {
+    return value != other.value;
+  }
+};
+
+struct InheritedConstIterator : BaseIntIterator {
+  using iterator_category = std::input_iterator_tag;
+  using pointer = int *;
+  using difference_type = int;
+
+  InheritedConstIterator(int value) : BaseIntIterator(value) {}
+  InheritedConstIterator(const InheritedConstIterator &other) = default;
+
+  InheritedConstIterator &operator++() {
+    value++;
+    return *this;
+  }
+  InheritedConstIterator operator++(int) {
+    auto tmp = InheritedConstIterator(value);
+    value++;
+    return tmp;
+  }
+};
+
+// MARK: Templated iterator types that use inheritance
+
+template <typename T>
+struct BaseTemplatedIterator {
+  using value_type = T;
+  using reference = const T &;
+
+  T value;
+
+  BaseTemplatedIterator(T value) : value(value) {}
+
+  reference operator*() const { return value; }
+
+  bool operator==(const BaseTemplatedIterator<T> &other) const {
+    return value == other.value;
+  }
+  bool operator!=(const BaseTemplatedIterator<T> &other) const {
+    return value != other.value;
+  }
+};
+
+template <typename T>
+struct InheritedTemplatedConstIterator : BaseTemplatedIterator<T> {
+  using iterator_category = std::input_iterator_tag;
+  using pointer = int *;
+  using difference_type = int;
+
+  InheritedTemplatedConstIterator(T value) : BaseTemplatedIterator<T>(value) {}
+  InheritedTemplatedConstIterator(const InheritedTemplatedConstIterator<T> &other) = default;
+
+  InheritedTemplatedConstIterator<T> &operator++() {
+    BaseTemplatedIterator<T>::value++;
+    return *this;
+  }
+  InheritedTemplatedConstIterator<T> operator++(int) {
+    auto tmp = InheritedTemplatedConstIterator<T>(BaseTemplatedIterator<T>::value);
+    BaseTemplatedIterator<T>::value++;
+    return tmp;
+  }
+};
+
+typedef InheritedTemplatedConstIterator<int> InheritedTemplatedConstIteratorInt;
+
+template <typename T>
+struct BaseTemplatedRACIterator {
+  using value_type = T;
+  using reference = const T &;
+  using difference_type = int;
+
+  T value;
+
+  BaseTemplatedRACIterator(T value) : value(value) {}
+
+  reference operator*() const { return value; }
+
+  bool operator==(const BaseTemplatedRACIterator<T> &other) const {
+    return value == other.value;
+  }
+  bool operator!=(const BaseTemplatedRACIterator<T> &other) const {
+    return value != other.value;
+  }
+};
+
+template <typename T>
+struct InheritedTemplatedConstRACIterator : BaseTemplatedRACIterator<T> {
+  using _super = BaseTemplatedRACIterator<T>;
+  using iterator_category = std::random_access_iterator_tag;
+  using pointer = int *;
+
+  InheritedTemplatedConstRACIterator(T value)
+      : BaseTemplatedRACIterator<T>(value) {}
+  InheritedTemplatedConstRACIterator(
+      const InheritedTemplatedConstRACIterator<T> &other) = default;
+
+  InheritedTemplatedConstRACIterator<T> &operator++() {
+    _super::value++;
+    return *this;
+  }
+  InheritedTemplatedConstRACIterator<T> operator++(int) {
+    auto tmp = InheritedTemplatedConstRACIterator<T>(_super::value);
+    _super::value++;
+    return tmp;
+  }
+
+  InheritedTemplatedConstRACIterator<T>
+  operator+(typename _super::difference_type v) const {
+    return {_super::value + v};
+  }
+  InheritedTemplatedConstRACIterator<T>
+  operator-(typename _super::difference_type v) const {
+    return {_super::value - v};
+  }
+  friend InheritedTemplatedConstRACIterator<T>
+  operator+(typename _super::difference_type v,
+            const InheritedTemplatedConstRACIterator<T> &it) {
+    return it + v;
+  }
+  int operator-(const InheritedTemplatedConstRACIterator<T> &other) const {
+    return _super::value - other.value;
+  }
+
+  void operator+=(typename _super::difference_type v) { _super::value += v; }
+  void operator-=(typename _super::difference_type v) { _super::value -= v; }
+
+  bool operator<(const InheritedTemplatedConstRACIterator<T> &other) const {
+    return _super::value < other.value;
+  }
+};
+
+typedef InheritedTemplatedConstRACIterator<int> InheritedTemplatedConstRACIteratorInt;
+
+template <typename T>
+struct BaseTemplatedRACIteratorOutOfLineOps {
+  using value_type = T;
+  using reference = const T &;
+  using difference_type = int;
+  
+  T value;
+
+  BaseTemplatedRACIteratorOutOfLineOps(T value) : value(value) {}
+
+  reference operator*() const { return value; }
+};
+
+template <typename T>
+bool operator==(const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+                const BaseTemplatedRACIteratorOutOfLineOps<T> &rhs) {
+  return lhs.value == rhs.value;
+}
+
+template <typename T>
+bool operator!=(const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+                const BaseTemplatedRACIteratorOutOfLineOps<T> &rhs) {
+  return lhs.value != rhs.value;
+}
+
+template <typename T>
+bool operator<(const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+               const BaseTemplatedRACIteratorOutOfLineOps<T> &rhs) {
+  return lhs.value < rhs.value;
+}
+
+template <typename T>
+typename BaseTemplatedRACIteratorOutOfLineOps<T>::difference_type
+operator-(const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+          const BaseTemplatedRACIteratorOutOfLineOps<T> &rhs) {
+  return lhs.value - rhs.value;
+}
+
+template <typename T>
+BaseTemplatedRACIteratorOutOfLineOps<T> operator+(
+    const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+    typename BaseTemplatedRACIteratorOutOfLineOps<T>::difference_type rhs) {
+  return {lhs.value + rhs};
+}
+
+template <typename T>
+BaseTemplatedRACIteratorOutOfLineOps<T> operator-(
+    const BaseTemplatedRACIteratorOutOfLineOps<T> &lhs,
+    typename BaseTemplatedRACIteratorOutOfLineOps<T>::difference_type rhs) {
+  return {lhs.value - rhs};
+}
+
+template <typename T>
+BaseTemplatedRACIteratorOutOfLineOps<T>
+operator+(typename BaseTemplatedRACIteratorOutOfLineOps<T>::difference_type lhs,
+          const BaseTemplatedRACIteratorOutOfLineOps<T> &rhs) {
+  return {rhs.value + lhs};
+}
+
+template <typename T>
+struct InheritedTemplatedConstRACIteratorOutOfLineOps
+    : BaseTemplatedRACIteratorOutOfLineOps<T> {
+  using _super = BaseTemplatedRACIteratorOutOfLineOps<T>;
+  using _self = InheritedTemplatedConstRACIteratorOutOfLineOps<T>;
+  using iterator_category = std::random_access_iterator_tag;
+  using pointer = int *;
+
+  InheritedTemplatedConstRACIteratorOutOfLineOps(T value) : _super(value) {}
+  InheritedTemplatedConstRACIteratorOutOfLineOps(const _self &other) = default;
+
+  _self &operator++() {
+    _super::value++;
+    return *this;
+  }
+  _self operator++(int) {
+    auto tmp = _self(_super::value);
+    _super::value++;
+    return tmp;
+  }
+
+  void operator+=(typename _super::difference_type v) { _super::value += v; }
+  void operator-=(typename _super::difference_type v) { _super::value -= v; }
+};
+
+typedef InheritedTemplatedConstRACIteratorOutOfLineOps<int>
+    InheritedTemplatedConstRACIteratorOutOfLineOpsInt;
+
 #endif // TEST_INTEROP_CXX_STDLIB_INPUTS_CUSTOM_ITERATOR_H

--- a/test/Interop/Cxx/stdlib/overlay/Inputs/custom-sequence.h
+++ b/test/Interop/Cxx/stdlib/overlay/Inputs/custom-sequence.h
@@ -108,4 +108,31 @@ struct HasTemplatedIterator {
 
 typedef HasTemplatedIterator<int> HasUninstantiatableIterator;
 
+struct HasInheritedConstIterator {
+private:
+  InheritedConstIterator b = InheritedConstIterator(1);
+  InheritedConstIterator e = InheritedConstIterator(6);
+
+public:
+  InheritedConstIterator begin() const { return b; }
+  InheritedConstIterator end() const { return e; }
+};
+
+template <typename T>
+struct HasInheritedTemplatedConstIterator {
+public:
+  typedef InheritedTemplatedConstIterator<int> iterator;
+
+private:
+  iterator b = iterator(1);
+  iterator e = iterator(7);
+
+public:
+  iterator begin() const { return b; }
+  iterator end() const { return e; }
+};
+
+typedef HasInheritedTemplatedConstIterator<int>
+    HasInheritedTemplatedConstIteratorInt;
+
 #endif // TEST_INTEROP_CXX_STDLIB_INPUTS_CUSTOM_SEQUENCE_H

--- a/test/Interop/Cxx/stdlib/overlay/custom-collection-module-interface.swift
+++ b/test/Interop/Cxx/stdlib/overlay/custom-collection-module-interface.swift
@@ -17,3 +17,15 @@
 // CHECK:   typealias Iterator = CxxIterator<SimpleCollectionReadOnly>
 // CHECK:   typealias RawIterator = SimpleCollectionReadOnly.iterator
 // CHECK: }
+
+// CHECK: struct HasInheritedTemplatedConstRACIterator<Int32> : CxxRandomAccessCollection {
+// CHECK:   typealias Element = InheritedTemplatedConstRACIterator<Int32>.Pointee
+// CHECK:   typealias Iterator = CxxIterator<HasInheritedTemplatedConstRACIterator<Int32>>
+// CHECK:   typealias RawIterator = InheritedTemplatedConstRACIterator<Int32>
+// CHECK: }
+
+// CHECK: struct HasInheritedTemplatedConstRACIteratorOutOfLineOps<Int32> : CxxRandomAccessCollection {
+// CHECK:   typealias Element = InheritedTemplatedConstRACIteratorOutOfLineOps<Int32>.Pointee
+// CHECK:   typealias Iterator = CxxIterator<HasInheritedTemplatedConstRACIteratorOutOfLineOps<Int32>>
+// CHECK:   typealias RawIterator = InheritedTemplatedConstRACIteratorOutOfLineOps<Int32>
+// CHECK: }

--- a/test/Interop/Cxx/stdlib/overlay/custom-collection.swift
+++ b/test/Interop/Cxx/stdlib/overlay/custom-collection.swift
@@ -49,4 +49,22 @@ CxxCollectionTestSuite.test("SimpleArrayWrapper as Swift.Collection") {
   expectEqual(mapped.last, 51)
 }
 
+CxxCollectionTestSuite.test("HasInheritedTemplatedConstRACIterator as Swift.Collection") {
+  let c = HasInheritedTemplatedConstRACIteratorInt()
+  expectEqual(c.first, 1)
+  expectEqual(c.last, 5)
+
+  let reduced = c.reduce(0, +)
+  expectEqual(reduced, 15)
+}
+
+CxxCollectionTestSuite.test("HasInheritedTemplatedConstRACIteratorOutOfLineOps as Swift.Collection") {
+  let c = HasInheritedTemplatedConstRACIteratorOutOfLineOpsInt()
+  expectEqual(c.first, 1)
+  expectEqual(c.last, 3)
+
+  let reduced = c.reduce(0, +)
+  expectEqual(reduced, 6)
+}
+
 runAllTests()

--- a/test/Interop/Cxx/stdlib/overlay/custom-convertible-to-collection.swift
+++ b/test/Interop/Cxx/stdlib/overlay/custom-convertible-to-collection.swift
@@ -59,4 +59,16 @@ CxxSequenceTestSuite.test("SimpleSequence.forEach") {
   expectEqual([1, 2, 3, 4] as [Int32], array)
 }
 
+CxxSequenceTestSuite.test("HasInheritedConstIterator to Swift.Array") {
+  let seq = HasInheritedConstIterator()
+  let array = Array(seq)
+  expectEqual([1, 2, 3, 4, 5] as [Int32], array)
+}
+
+CxxSequenceTestSuite.test("HasInheritedTemplatedConstIterator to Swift.Array") {
+  let seq = HasInheritedTemplatedConstIteratorInt()
+  let array = Array(seq)
+  expectEqual([1, 2, 3, 4, 5, 6] as [Int32], array)
+}
+
 runAllTests()

--- a/test/Interop/Cxx/stdlib/overlay/custom-iterator-module-interface.swift
+++ b/test/Interop/Cxx/stdlib/overlay/custom-iterator-module-interface.swift
@@ -105,3 +105,23 @@
 // CHECK:   typealias Pointee = Int32
 // CHECK:   typealias Distance = TemplatedRACIteratorOutOfLineEq<Int32>.difference_type
 // CHECK: }
+
+// CHECK: struct BaseIntIterator {
+// CHECK: }
+// CHECK: struct InheritedConstIterator : UnsafeCxxInputIterator {
+// CHECK: }
+
+// CHECK: struct InheritedTemplatedConstIterator<T> {
+// CHECK: }
+// CHECK: struct InheritedTemplatedConstIterator<Int32> : UnsafeCxxInputIterator {
+// CHECK: }
+
+// CHECK: struct InheritedTemplatedConstRACIterator<T> {
+// CHECK: }
+// CHECK: struct InheritedTemplatedConstRACIterator<Int32> : UnsafeCxxRandomAccessIterator, UnsafeCxxInputIterator {
+// CHECK: }
+
+// CHECK: struct InheritedTemplatedConstRACIteratorOutOfLineOps<T> {
+// CHECK: }
+// CHECK: struct InheritedTemplatedConstRACIteratorOutOfLineOps<Int32> : UnsafeCxxRandomAccessIterator, UnsafeCxxInputIterator {
+// CHECK: }

--- a/test/Interop/Cxx/stdlib/use-std-map.swift
+++ b/test/Interop/Cxx/stdlib/use-std-map.swift
@@ -29,7 +29,6 @@ StdMapTestSuite.test("Map.subscript") {
   expectNil(m[5])
 }
 
-#if !os(Linux) // TODO: enable on Linux (rdar://105220600)
 StdMapTestSuite.test("UnorderedMap.subscript") {
   // This relies on the `std::unordered_map` conformance to `CxxDictionary` protocol.
   var m = initUnorderedMap()
@@ -39,6 +38,5 @@ StdMapTestSuite.test("UnorderedMap.subscript") {
   expectNil(m[-1])
   expectNil(m[5])
 }
-#endif
 
 runAllTests()

